src/deploy/functions/iambinder.ts

```diff
@@ -0,0 +1,286 @@
+        "roles/secretmanager.secretAccessor",
+        [serviceAccount]
+      );
+    }
+
+    if (backend.isEventTriggered(endpoint)) {
+      if (endpoint.platform === "gcfv2") {
+        // Eventarc requires pubsub service account to have serviceAccountTokenCreator role.
+        // See https://cloud.google.com/eventarc/docs/roles-permissions#triggertypes-roles.
+        this.addBinding(
+          `//cloudresourcemanager.googleapis.com/projects/${this.projectNumber}`,
+          "roles/iam.serviceAccountTokenCreator",
+          [`serviceAccount:service-${this.projectNumber}@gcp-sa-pubsub.iam.gserviceaccount.com`]
+        );
+        // Service account associated w/ eventarc trigger must have permission to invoke a Cloud Run instance.
+        // We don't allow users to specify this service account hence it will always be the default compute service
+        // account. Usually, the default service account has EDITOR role which includes permission to invoke a Cloud
+        // Run instance, but given strict org policy it may not have this permission.
+        // See https://cloud.google.com/compute/docs/access/service-accounts#default_service_account.
+        this.addBinding(
+          `//cloudresourcemanager.googleapis.com/projects/${this.projectNumber}`,
+          "roles/run.invoker",
+          [this.defaultComputeServiceAccount(endpoint)]
+        );
+        if (v2Events.isStorageTriggered(endpoint)) {
+          // Storage account associated w/ Cloud Storage must have permission to publish pubsub messages.
+          this.addBinding(
+            `//cloudresourcemanager.googleapis.com/projects/${this.projectNumber}`,
+            "roles/pubsub.publisher",
+            [await this.fetchStorageServiceAccount()]
+          );
+        }
+      }
+    }
+  }
+
+  private addBinding(resource: Resource, role: Role, members: string[]): void {
+    const bindings = this.additions[resource] || new IamBindings();
+    bindings.add(role, members);
+    this.additions[resource] = bindings;
+  }
+
+  private async fetchStorageServiceAccount(): Promise<string> {
+    if (!this.storageServiceAccount) {
+      const resp = await storage.getServiceAccount(this.projectId);
+      this.storageServiceAccount = `serviceAccount:${resp.email_address}`;
+    }
+    return this.storageServiceAccount;
+  }
+
+  private defaultComputeServiceAccount(endpoint: backend.Endpoint): string {
+    if (endpoint.platform === "gcfv1") {
+      return `serviceAccount:${this.projectId}@appspot.gserviceaccount.com`;
+    } else if (endpoint.platform === "gcfv2") {
+      return `serviceAccount:${this.projectNumber}-compute@developer.gserviceaccount.com`;
+    }
+    assertExhaustive(endpoint.platform);
+  }
+
+  static async updatePolicy(resource: Resource, bindings: IamBindings): Promise<iam.Policy> {
+    const existingPolicy = await IamBinder.getPolicy(resource);
+    const existingBindings = IamBindings.fromIamBindings(existingPolicy.bindings || []);
```

Comment:
Nit. If this is the design pattern you'll always use you can probably just make `getPolicy` return an `IamBindings`